feed-formats/buttwoo.js

```diff
@@ -0,0 +1,361 @@
+          contentHash,
+        ]
+        const encodedValue = bipf.allocAndEncode(value)
+        const sigBuf = base64ToBuffer(signature)
+        return bipf.allocAndEncode([encodedValue, sigBuf, contentBuffer])
+      } else {
+        // prettier-ignore
+        throw new Error(`Feed format "${feedFormat.name}" does not support encoding "${encoding}"`)
+      }
+    },
+
+    validateSingle: (hmacKey, nativeMsg, prevNativeMsg, cb) => {
+      const [encodedValue, signature, contentBuffer] = bipf.decode(nativeMsg)
+      const [
+        authorBFE,
+        parentBFE,
+        sequence,
+        timestamp,
+        previousBFE,
+        tag,
+        contentSize,
+        contentHash,
+      ] = bipf.decode(encodedValue)
+
+      if (contentHash.length !== 32)
+        return cb(new Error('Content hash wrong size: ' + contentHash.length))
+
+      if (!Buffer.isBuffer(tag))
+        return cb(new Error('Tag is not a buffer: ' + tag))
+      if (tag.length !== 1)
+        return cb(new Error('Tag is not a single byte: ' + tag))
+      const byte = tag[0]
+      if (byte < 0 || byte > 2)
+        return cb(new Error('Tag is not valid: ' + byte))
+
+      if (contentBuffer.length !== contentSize)
+        return cb(new Error('Content size does not match content'))
+
+      const testedContentHash = blake3.hash(contentBuffer)
+      if (Buffer.compare(testedContentHash, contentHash) !== 0)
+        return cb(new Error('Content hash does not match content'))
+
+      if (
+        typeof timestamp !== 'number' ||
+        isNaN(timestamp) ||
+        !isFinite(timestamp)
+      )
+        return cb(
+          new Error(
+            `invalid message: timestamp is "${timestamp}", expected a JavaScript number`
+          )
+        )
+
+      const author = bfe.decode(authorBFE)
+      const { data: public } = SSBURI.decompose(author)
+      const key = { public, curve: 'ed25519' }
+
+      if (!ssbKeys.verify(key, signature, hmacKey, encodedValue))
+        return cb(new Error('Signature does not match encoded value'))
+
+      // FIXME: check correct BFE types!
+      // FIXME: check length of content
+
+      if (prevNativeMsg !== null) {
+        const [encodedValuePrev] = bipf.decode(prevNativeMsg)
+        const [
+          authorBFEPrev,
+          parentBFEPrev,
+          sequencePrev,
+          timestampPrev,
+          previousBFEPrev,
+          tagPrev,
+        ] = bipf.decode(encodedValuePrev)
+
+        if (Buffer.compare(authorBFE, authorBFEPrev) !== 0)
+          return cb(new Error('Author does not match previous message'))
+
+        if (Buffer.compare(parentBFE, parentBFEPrev) !== 0)
+          return cb(new Error('Parent does not match previous message'))
+
+        if (sequence !== sequencePrev + 1)
+          return cb(new Error('Sequence must increase'))
+
+        if (timestamp <= timestampPrev)
+          return cb(new Error('Timestamp must increase'))
+
+        if (Buffer.compare(previousBFE, previousKeyBFE) !== 0)
```

Comment:
`previousKeyBFE` is not defined here. I'll push up a counter PR with some minor fixes to this file so please wait for that :)

Reply:
I fixed this in a recent commit, it should be orthogonal to https://github.com/ssbc/ssb-db2/pull/348